### PR TITLE
feat(sso): SSO enforcement

### PR DESCRIPTION
**Pull request by paolodamico**
## Problem

Part of the full SAML support https://github.com/PostHog/posthog/issues/5647 & SSO enforcement on Cloud (compliance feature).

## Changes

**This PR:**
- Fully enables SSO enforcement via authentication domains. User won’t be able to log in with password or request password reset if SSO is enforced for their domain.
- Removes SSO enforcement via env var (for self-hosted) in favor of everyone enforcing SSO the same way (see details below).
- Enables the login precheck for everyone.

**Advantages:**
- Less confusion around how to configure SSO enforcement. There’s just one place, and it’s easy to manage in the app.
- More maintainable code and lower security risk. SSO enforcement in a single place.

**Downsides:**
- Users on a single tenant company on self-hosted that has SAML enforced will now have to type their email address or click on the SAML button (as opposed to automatic redirection).
- Users on PostHog FOSS where SSO enforcement wouldn’t be available will have to type their email address nonetheless.
- There’s currently an edge case where you could create an account in a self-hosted instance with an email address for which SSO is not enforced to bypass the requirement (will push an improvement for this).

## How did you test this code?
- Django functional tests.

### CODE REVIEW

**Comment by paolodamico** [comment]
If you're on board with this approach @mariusandra we should definitely merge before the release tomorrow, otherwise users will get a deprecation notice and instructions to update to something that is already "re-deprecated" 😂

**Comment by mariusandra** [comment]
Saw the request to review, and have been waiting for the "fix fix" queue to end 😅

File: posthog/models/organization_domain.py [hunk 1]
```diff
@@ -19,6 +21,40 @@ def verified_domains(self):
...
+
+        # Check SSO provider is properly configured
+        sso_providers = get_available_sso_providers()
+        if not sso_providers[candidate_sso_enforcement]:
+            print_warning(
+                [
+                    f"SSO is enforced for domain {domain} but the SSO provider ({candidate_sso_enforcement}) is not properly configured."
+                ]
+            )
+            return None
```
Comment (mariusandra) [inline posthog/models/organization_domain.py:54]: I don't know if this is the right place to log this data.

The [backend coding conventions](https://posthog.com/handbook/engineering/conventions/backend-coding#logging) don't speak of `print_warning`, but more of `structlog`. CC @guidoiaquinti

**Review by mariusandra — COMMENTED** [review]
Overall looks good, not sure what's up with tests though. 

Some comments inline, and also a question: If you only had an account via SSO and the EE license expires, can you somehow log in without a password? This may be paranoia, but what about adding another test case that makes sure you can't bypass the login if there's no password set, or something like that.

File: ee/api/test/test_authentication.py [hunk 2]
```diff
@@ -55,53 +56,143 @@
...
+            organization=self.organization,
+            verified_at=timezone.now(),
+            sso_enforcement="github",
+        )
+        User.objects.create_and_join(self.organization, "i_do_not_exist@anotherdomain.com", self.CONFIG_PASSWORD)
+
+        with self.settings(**GITHUB_MOCK_SETTINGS):
+            response = self.client.post("/api/login/precheck", {"email": "i_do_not_exist@anotherdomain.com"})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.json(), {"sso_enforcement": "github"})
```
[the remark this answers is not in the document]
Comment (paolodamico) [reply ee/api/test/test_authentication.py:116]: Yes, `test_login_precheck_with_enforced_sso` checks with an account that actually exists, `test_login_precheck_with_inexistent_account` checks with an account that doesn't exist but for which the domain still has enforcement rules. It basically helps prove that the actual email address is not important, but the domain.

**Comment by paolodamico** [comment]
Thanks for the review @mariusandra! Re the license expiring, you would still be unable to log in if there's no password set, but you could technically request a password reset there and log in with a password. Should we not allow this?

**Review by mariusandra — APPROVED** [review]
The [logging format](https://posthog.com/handbook/engineering/conventions/backend-coding#logging#format) is still not right. In case the docs are confusing as to what to do, any PRs are welcome 😅 . Other than that, looks good... and I assume you'll be testing it in every possible configuration for breaking the release?

**Comment by paolodamico** [comment]
Thanks @mariusandra, I'm doing everything so quickly that I'm making silly oversights. Yup we should test thoroughly in break the release.
